### PR TITLE
Tensorflow improvements

### PR DESCRIPTION
**Pull request by jplu**
Hello,

Here a quite big PR that propose the following updates:

- Loss computation is now attached to their respective class, such as PyTorch.
- Remove useless `mode` and `loss_name` parameters for the TF Trainer.
- Add missing task models to different Transformers
- Bugfix on T5 keras serialization + tests
- Add tests for TF Flaubert and XLM-Roberta
- Bugfix in TF Trainer for Tensorflow 2.2

Reviews are welcome :)

/cc @julien-c @LysandreJik @thomwolf

### CODE REVIEW

**Comment by jplu** [comment]
Some commits are missing... I think it is due to the high number of error rate from Github.

File: src/transformers/modeling_tf_auto.py [hunk 1]
```diff
@@ -654,11 +738,13 @@ def from_pretrained(cls, pretrained_model_name_or_path, *model_args, **kwargs):
...
         """
         config = kwargs.pop("config", None)
+
         if not isinstance(config, PretrainedConfig):
             config = AutoConfig.from_pretrained(pretrained_model_name_or_path, **kwargs)
 
         for config_class, model_class in TF_MODEL_WITH_LM_HEAD_MAPPING.items():
-            if isinstance(config, config_class):
+            # Not using isinstance() here to do not take into account inheritance
+            if config_class == type(config):
```
Comment (julien-c) [inline src/transformers/modeling_tf_auto.py:747]: in pytorch the different configs are sorted so that you never get a child class before its parent (precisely to prevent this), but this is reasonable solution too

File: src/transformers/modeling_tf_utils.py [hunk 2]
```diff
@@ -104,6 +106,41 @@ def get_config(self):
...
+            from_logits=True, reduction=tf.keras.losses.Reduction.NONE
+        )
+        active_loss = tf.reshape(labels, (-1,)) != -1
+        reduced_logits = tf.boolean_mask(tf.reshape(logits[0], (-1, shape_list(logits[0])[2])), active_loss)
+        labels = tf.boolean_mask(tf.reshape(labels, (-1,)), active_loss)
+
+        return loss_fn(labels, reduced_logits)
+
+
+class TFSequenceClassificationAndMultipleChoiceLoss:
```
Comment (julien-c) [inline src/transformers/modeling_tf_utils.py:132]: should we split into two different classes?

Comment (jplu) [reply src/transformers/modeling_tf_utils.py:132]: My thought was, as it is the exact same loss computation why not merge the two names, but your proposal might be more insightful indeed.

Comment (julien-c) [reply src/transformers/modeling_tf_utils.py:132]: Maybe just alias one to the other or do a trivial sub-class

Comment (jplu) [reply src/transformers/modeling_tf_utils.py:132]: Very good point, I will do the update

Comment (jplu) [reply src/transformers/modeling_tf_utils.py]: Should be ok now.

File: src/transformers/modeling_tf_albert.py [hunk 3]
```diff
@@ -1036,10 +1103,7 @@ def call(
             inputs_embeds = inputs[5] if len(inputs) > 5 else inputs_embeds
             assert len(inputs) <= 6, "Too many inputs."
         elif isinstance(inputs, dict):
-            print("isdict(1)")
             input_ids = inputs.get("input_ids")
-            print(input_ids)
-
```
Comment (LysandreJik) [inline src/transformers/modeling_tf_albert.py:1042]: Nice catch!

**Review by LysandreJik — COMMENTED** [review]
This is great work, love the added flexibility to the API and how it similar to our Pytorch model's API it can be. I like the coding style. 

I find this is a bit different to the PyTorch API as:

1) it uses Mixins, and I'm okay with them as I think the readability is still good, while it does add a (maybe unnecessary?) layer of abstraction. It does greatly improve code sharing across models though, which is welcome.
2) Loss isn't computed when passing labels, but by directly calling `model.compute_loss(x, y)`

I'm not opposed to the first point, but a bit more to the second point. Is there something that prevents using `labels` in TensorFlow as we do it in PyTorch? As we're aiming at API compatibility, I think this is something we should get right.

**Comment by jplu** [comment]
Thanks @LysandreJik for your constructive comments!

For the second point, before to answer in order to be sure, you mean that it would be more convenient that the output of the `call(...)` methods in the TF tasks model returns the same tuple `(loss), logits, (hidden_states), (attentions)` than the `forward(...)` methods in PT tasks model?

**Comment by LysandreJik** [comment]
Yes, that's what I mean. I think having this to be the same as the PyTorch API would make sense. It wouldn't be a breaking change either, as it would require the `labels` to be passed to the model.

I think doing this could still leverage Mixins, by calling a `self._compute_loss` or `self.compute_loss` if we want to expose this method as well. I have no strong opinion on that last item.

**Comment by jplu** [comment]
Ok, indeed makes sense and I don't think it is a problem to do that way, I will work on this today to see if there is any issue that would not allow us to do that.

**Comment by julien-c** [comment]
I agree with @LysandreJik's 2nd point – maybe we can even take advantage of this to implement named tuples for TF models output, like @thomwolf and @patrickvonplaten intend to do for PyTorch (as it's going to be a breaking change in TF models anyways, maybe we can do this at the same time?)

**Comment by jplu** [comment]
Since my last commit, now the TF models return the loss such as the PT ones if the labels are given. 

About the named tuples, looks to be a good idea indeed, but I think we should implement this in another PR in order to release this in same time than for PT. No?

**Comment by julien-c** [comment]
> About the named tuples [...] we should implement this in another PR in order to release this in same time than for PT. No?

Yes, makes sense!

**Comment by jplu** [comment]
Ok, looks good to me, I have tested the new models with different examples that use the trainer and they all work, tests looks to be ok as well except the quality one that I don't know how to fix :smile:

File: src/transformers/modeling_tf_t5.py [hunk 4]
```diff
@@ -25,7 +25,8 @@
 
 from .configuration_t5 import T5Config
 from .file_utils import DUMMY_INPUTS, DUMMY_MASK, add_start_docstrings, add_start_docstrings_to_callable
-from .modeling_tf_utils import TFPreTrainedModel, TFSharedEmbeddings, shape_list
+from .modeling_tf_utils import TFPreTrainedModel, TFSharedEmbeddings, keras_serializable, shape_list
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_t5.py:28]: Thanks for the changes here! Looks good to me

File: src/transformers/modeling_tf_transfo_xl.py [hunk 5]
```diff
@@ -734,7 +734,7 @@ def call(self, inputs, **kwargs):
         return outputs
 
 
-class TFTransfoXLLMHead(tf.keras.layers.Layer):
+class TFTransfoXLWithLMHeadModel(tf.keras.layers.Layer):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_transfo_xl.py:737]: This is a breaking change, no? Maybe we want to add an alias for `TFTransfoXLLMHead` for backward compatibility @LysandreJik

File: src/transformers/optimization_tf.py [hunk 6]
```diff
@@ -58,27 +58,40 @@ def get_config(self):
...
-        learning_rate=lr_schedule,
-        weight_decay_rate=0.01,
-        beta_1=0.9,
-        beta_2=0.999,
-        epsilon=1e-6,
-        exclude_from_weight_decay=["LayerNorm", "layer_norm", "bias"],
-    )
-
-    return optimizer
+    return optimizer, lr_schedule
```
Comment (patrickvonplaten) [inline src/transformers/optimization_tf.py:94]: This is also a breaking change - we should document this well so that the user knows

File: src/transformers/modeling_tf_transfo_xl.py (hunk 5 above)
Comment (jplu) [reply src/transformers/modeling_tf_transfo_xl.py]: It is my bad, I just done a commit to rename it.

File: src/transformers/optimization_tf.py (hunk 6 above)
Comment (jplu) [reply src/transformers/optimization_tf.py:95]: Done!

File: examples/benchmarking/plot_csv_file.py [hunk 7]
```diff
@@ -3,9 +3,9 @@
 from dataclasses import dataclass, field
 from typing import Optional
 
+import matplotlib.pyplot as plt
```
Comment (patrickvonplaten) [inline examples/benchmarking/plot_csv_file.py:6]: I had problems with isort on this file as well :D I think you might just reverse this change to fix isort

**Comment by patrickvonplaten** [comment]
A more general question regarding training in tensorflow in general (I'm not super familiar with TF 2.0 training, so I'm asking primarly to learn a bit :-) ): 
I remember that when TF 2.0 was not out, most people used keras to train a model with 
`model.fit(x_train, y_train)` => is this still the case?
or are people more and more switching to the TF 2.0 training style as shown here: https://www.tensorflow.org/tutorials/quickstart/advanced and which basically consists of using 
`optimizer.apply_gradients(zip(gradients, model.trainable_variables))`. This is also what we do in the TF trainer right? 

Was it possible and recommended to train transformer models with keras' `model.train()` before TF Trainer and is it still possible now?

**Comment by jplu** [comment]
This is a good question! Short answer: yes it is still possible but witthout any gradient accumulation, that's mostly why the trainer uses the advanced training of TensorFlow.

I'm currently preparing a next PR that will integrate the new `Model.train_step` feature added in [TF 2.2](https://github.com/tensorflow/tensorflow/releases/tag/v2.2.0). Basically this update allows you to create your own train step, and then integrate the missing gradient accumulation but this new PR will be only for TF >= 2.2.

File: src/transformers/optimization_tf.py (hunk 6 above)
Comment (LysandreJik) [reply src/transformers/optimization_tf.py:95]: Indeed, this is not backwards compatible. @jplu, do you expect this method to be currently used by users outside of the `Trainer`? Would this breaking change impact those users?

Comment (jplu) [reply src/transformers/optimization_tf.py:95]: Honestly no, I not expecting users to use this outside the TF Trainer. Also the TF trainer has been updated to use this new return format, such as the PT one.

Comment (LysandreJik) [reply src/transformers/optimization_tf.py:95]: Right, I thought so too!

File: examples/benchmarking/plot_csv_file.py (hunk 7 above)
Comment (jplu) [reply examples/benchmarking/plot_csv_file.py]: Done ! And thanks for the hint :smile:

**Comment by LysandreJik** [comment]
@patrickvonplaten It was possible and we definitely aim to keep compatibility with keras' `fit` method. We don't have many tutorials that cover it, though, having some would probably make it easier for new users coming from Keras to use our lib.

@julien-c, we've had the offline approval from @thomwolf, feel free to merge when you want. Glad to welcome this in the library!

**Comment by julien-c** [comment]
Just tweaked the training_args.logging_dir to keep the same default as pytorch (I like that it creates a new subfolder each time you relaunch a training)

Great job @jplu, thank you 💪
